algebra/localization.lean

```diff
@@ -86,6 +97,169 @@ by refine
+    is_ideal.univ_of_one_mem P h,
+  mul_mem := λ x y hnx hny hxy, or.cases_on
+    (is_prime_ideal.mem_or_mem_of_mul_mem hxy) hnx hny }
+
+def at_prime := loc α (set.compl P)
+
+instance at_prime.local_ring :
+  @local_ring (at_prime P) (loc.comm_ring α _) :=
+local_of_nonunits_ideal
+  (λ hze, have _, from quotient.exact hze, let ⟨t, hts, ht⟩ := this in
+     hts $ have htz : t = 0, by simpa using ht,
+       suffices (0:α) ∈ P, by rwa htz,
+       is_ideal.zero_mem P)
+  (λ x y hx hy ⟨z, hz⟩,
+     let ⟨⟨r₁, s₁, hs₁⟩, hrs₁⟩ := quotient.exists_rep x,
+         ⟨⟨r₂, s₂, hs₂⟩, hrs₂⟩ := quotient.exists_rep y,
+         ⟨⟨r₃, s₃, hs₃⟩, hrs₃⟩ := quotient.exists_rep z in
+     have _, by rw [← hrs₁, ← hrs₂, ← hrs₃] at hz; from quotient.exact hz,
+     let ⟨t, hts, ht⟩ := this in
+     have hr₁ : r₁ ∈ P, from classical.by_contradiction $
+     λ hnr₁, hx ⟨⟦⟨s₁, r₁, hnr₁⟩⟧, by rw ←hrs₁;
+       from quotient.sound ⟨1, is_submonoid.one_mem _, by simp [mul_comm]⟩⟩,
+     have hr₂ : r₂ ∈ P, from classical.by_contradiction $
+     λ hnr₂, hy ⟨⟦⟨s₂, r₂, hnr₂⟩⟧, by rw ←hrs₂;
+       from quotient.sound ⟨1, is_submonoid.one_mem _, by simp [mul_comm]⟩⟩,
+     or.cases_on (mem_or_mem_of_mul_eq_zero P ht) hts $
+     λ h, have h1 : (r₁ * s₂ + r₂ * s₁) * r₃ - s₁ * s₂ * s₃ ∈ P,
+     by simpa using h,
+     have h2 : (r₁ * s₂ + r₂ * s₁) * r₃ ∈ P,
+     from is_ideal.mul_mem $ is_ideal.add_mem
+         (is_ideal.mul_mem hr₁)
+         (is_ideal.mul_mem hr₂),
+     or.cases_on (is_prime_ideal.mem_or_mem_of_mul_mem $
+         is_ideal.mem_of_mem_of_sub_mem h2 h1)
+       (λ h3, or.cases_on (is_prime_ideal.mem_or_mem_of_mul_mem h3) hs₁ hs₂)
+       hs₃)
+
+end at_prime
+
+def closure (S : set α) : set α := {y | ∃ (L:list α) (H:∀ x ∈ L, x ∈ S), L.prod = y }
+
+instance closure.is_submonoid (S : set α) : is_submonoid α (closure S) :=
+{ one_mem := ⟨[], by simp⟩,
+  mul_mem := λ x₁ x₂ ⟨L₁, hLS₁, hL₁⟩ ⟨L₂, hLS₂, hL₂⟩,
+    ⟨L₁ ++ L₂,
+     λ x hx, or.cases_on (list.mem_append.1 hx) (hLS₁ x) (hLS₂ x),
+     by simp [list.prod_append, *]⟩ }
+
+variable α
+
+def non_zero_divisors : set α := {x | ∀ z, x * z = 0 → z = 0}
+
+instance non_zero_divisors.is_submonoid : is_submonoid α (non_zero_divisors α) :=
+{ one_mem := λ z hz, by simpa using hz,
+  mul_mem := λ x₁ x₂ hx₁ hx₂ z hz,
+    have x₁ * (x₂ * z) = 0,
+    by rwa mul_assoc at hz,
+    hx₂ z $ hx₁ (x₂ * z) this }
+
+def quotient_ring := loc α (non_zero_divisors α)
+
+section quotient_ring
+
+variables {β : Type u} [integral_domain β] [decidable_eq β]
+
+lemma ne_zero_of_mem_non_zero_divisors {x : β} :
+  x ∈ loc.non_zero_divisors β → x ≠ 0 :=
+λ hm hz, have x * 1 = 0, by simp [hz], zero_ne_one (hm 1 this).symm
+
+lemma eq_zero_of_ne_zero_of_mul_eq_zero {x y : β} : x ≠ 0 → x * y = 0 → y = 0 :=
+λ hnx hxy, match eq_zero_or_eq_zero_of_mul_eq_zero hxy with
+| or.inl hx := false.elim $ hnx hx
+| or.inr hy := hy
+end
+
+lemma mem_non_zero_divisors_of_ne_zero {x : β} :
+  x ≠ 0 → x ∈ loc.non_zero_divisors β :=
```

Comment:
You should combine these three theorems into one biconditional. (The second one is just `or.resolve_left` applied to `mul_eq_zero`.)